pkg/clientgocred/cred.go

```diff
@@ -0,0 +1,370 @@
+)
+
+type credential struct {
+	AccessToken string                 `json:"access_token"`
+	TokenExpiry time.Time              `json:"token_expiry"`
+	X           map[string]interface{} `json:"-"` // Rest of the fields should go here.
+}
+
+// gcloudConfiguration is the struct unmarshalled
+// from gcloud config in json format
+type gcloudConfiguration struct {
+	Credential credential             `json:"credential"`
+	X          map[string]interface{} `json:"-"` // Rest of the fields should go here.
+}
+
+// cache is the struct that gets cached in the cache file in json format.
+// {
+//    "current_context": "gke_user-gke-dev_us-central1_autopilot-cluster-11",
+//    "access_token": "ya29.A0ARrdaM8WL....G0xYXGIQNPi5WvHe07ia4Gs",
+//    "token_expiry": "2022-01-27T08:27:52Z"
+// }
+type cache struct {
+	// CurrentContext refers to which context the token was last retrieved for. If
+	// currentContext in kubeconfig is changed, the current cached access token is invalidated.
+	CurrentContext string `json:"current_context"`
+	// AccessToken is gcloud access token
+	AccessToken string `json:"access_token"`
+	// TokenExpiry is gcloud access token's expiry.
+	TokenExpiry string `json:"token_expiry"`
+}
+
+// pluginContext holds data to be passed around (eg: useApplicationDefaultCredentials)
+// as well as methods that may needs to be mocked in test scenarios.
+type pluginContext struct {
+	googleDefaultTokenSource         func(ctx context.Context, scope ...string) (oauth2.TokenSource, error)
+	gcloudConfigOutput               func() ([]byte, error)
+	k8sStartingConfig                func() (*clientcmdapi.Config, error)
+	cachedToken                      func(pc *pluginContext) (string, string)
+	writeCacheFile                   func(content string) error
+	useApplicationDefaultCredentials bool
+}
+
+func newPluginContext(options *Options) *pluginContext {
+	return &pluginContext{
+		googleDefaultTokenSource:         google.DefaultTokenSource,
+		k8sStartingConfig:                k8sStartingConfig,
+		gcloudConfigOutput:               gcloudConfigOutput,
+		cachedToken:                      cachedToken,
+		writeCacheFile:                   writeCacheFile,
+		useApplicationDefaultCredentials: options.UseApplicationDefaultCredentials,
+	}
+}
+
+// Options struct inputs to PrintCred
+type Options struct {
+	UseApplicationDefaultCredentials bool
+}
+
+// PrintCred prints ExecCredentail to stdout  to be consumed by kubectl to connect to GKE Clusters
+// {
+//    "kind": "ExecCredential",
+//    "apiVersion": "client.authentication.k8s.io/v1beta1",
+//    "spec": {
+//        "interactive": false
+//    },
+//    "status": {
+//        "expirationTimestamp": "2022-01-27T07:10:46Z",
+//        "token": "ya29.A0ARrda.......0jDi8weH-36jJNru6Ps"
+//    }
+// }
+func PrintCred(options *Options) error {
+	if options == nil {
+		options = &Options{
+			UseApplicationDefaultCredentials: false,
+		}
+	}
+	pc := newPluginContext(options)
```

Comment:
Leaving this till the end.. Mostly to avoid code churn. 
My very first iteration was having a plugin struct and giving it some methods (pointer receivers). Then I thought, pluginContext was the preferred approach. So, I switched it to this. I will switch it back in the very end.